src/Numerics/LinearSolvers/BatchedGeneralizedMinimalResidualSolver.jl

```diff
@@ -0,0 +1,824 @@
+        ArrayType = ArrayType,
+        reshape_tuple_f = reshaping_tup,
+        permute_tuple_f = permute_tuple_f,
+    )
+end
+
+# initialize function (1)
+function LS.initialize!(
+    linearoperator!,
+    Q,
+    Qrhs,
+    solver::BatchedGeneralizedMinimalResidual,
+    args...,
+)
+    # body of initialize function in abstract iterative solver
+    return false, zero(eltype(Q))
+end
+
+# iteration function (2)
+function LS.doiteration!(
+    linearoperator!,
+    Q,
+    Qrhs,
+    gmres::BatchedGeneralizedMinimalResidual,
+    threshold,
+    args...,
+)
+    # initialize gmres.x
+    convert_structure!(gmres.x, Q, gmres.reshape_tuple_f, gmres.permute_tuple_f)
+    # apply linear operator to construct residual
+    r_vector = copy(Q)
+    linearoperator!(r_vector, Q, args...)
+    @. r_vector = Qrhs - r_vector
+    # The following ar and rr are technically not correct in general cases
+    ar = norm(r_vector)
+    rr = norm(r_vector) / norm(Qrhs)
+    # check if the initial guess is fantastic
+    if (ar < gmres.atol) || (rr < gmres.rtol)
+        return true, 0, ar
+    end
+    # initialize gmres.b
+    convert_structure!(
+        gmres.b,
+        r_vector,
+        gmres.reshape_tuple_f,
+        gmres.permute_tuple_f,
+    )
+    # apply linear operator to construct second krylov vector
+    linearoperator!(Q, r_vector, args...)
+    # initialize gmres.sol
+    convert_structure!(
+        gmres.sol,
+        Q,
+        gmres.reshape_tuple_f,
+        gmres.permute_tuple_f,
+    )
+    # initialize the rest of gmres
+    event = initialize_gmres!(gmres)
+    wait(event)
+    ar, rr = compute_residuals(gmres, 1)
+    # check if converged
+    if (ar < gmres.atol) || (rr < gmres.rtol)
+        event = construct_solution!(1, gmres)
+        wait(event)
+        convert_structure!(
+            Q,
+            gmres.x,
+            gmres.reshape_tuple_b,
+            gmres.permute_tuple_b,
+        )
+        return true, 1, ar
+    end
+    # body of iteration
+    @inbounds for i in 2:(gmres.k_n)
+        convert_structure!(
+            r_vector,
+            view(gmres.Q, :, i, :),
+            gmres.reshape_tuple_b,
+            gmres.permute_tuple_b,
+        )
+        linearoperator!(Q, r_vector, args...)
+        convert_structure!(
+            gmres.sol,
+            Q,
+            gmres.reshape_tuple_f,
+            gmres.permute_tuple_f,
+        )
+        event = gmres_update!(i, gmres)
+        wait(event)
+        ar, rr = compute_residuals(gmres, i)
+        # check if converged
+        if (ar < gmres.atol) || (rr < gmres.rtol)
+            event = construct_solution!(i, gmres)
+            wait(event)
+            convert_structure!(
+                Q,
+                gmres.x,
+                gmres.reshape_tuple_b,
+                gmres.permute_tuple_b,
+            )
+            return true, i, ar
+        end
+    end
+
+    event = construct_solution!(gmres.k_n, gmres)
+    wait(event)
```

Comment:
Can we get away with a device wait here? Either `wait(CPU(), event)` or `wait(GPU(), event)`.